### PR TITLE
WIP: fix a few obvious errors in aea_projection.py

### PR DESCRIPTION
**Pull request by rainwoodman**
It at least doesn't immediately crash.

### CODE REVIEW

**Comment by rainwoodman** [comment]
Currently it plots RA and DEC more or less correctly.
Here is an example:

```
import aea_projection
figure(figsize=(12, 8))

ra = random.uniform(size=10000, low=0, high=360)
dec = random.uniform(size=10000, low=-90, high=90)

a = subplot(111, projection="aea", dec_1=0., dec_2=60., ra_0=0., dec_0=0.)
a.set_latitude_grid(30)
a.set_longitude_grid_ends(60)
ra = random.uniform(size=10000, low=0, high=360)
dec = random.uniform(size=10000, low=-90, high=90)
grid()
p = plot((-30, -30), (-90, 90), 'x-')
plot(ra, dec, '. ')
axhline(30, )
axvline(30)
#p = axhline(y=60)
#grid()      
```

**Comment by pmelchior** [comment]
Thanks for you help! Have you done this before? I felt rather lost in the details of matplotlib.Axes.
Anyway, it seems you're still working on it. Let me know when you're done with it, and I'll review it.

**Comment by rainwoodman** [comment]
I am next to giving it up... Look at this:

index

**Comment by rainwoodman** [comment]
and this:

index

**Comment by pmelchior** [comment]
The first one is weird, the second one looks pretty reasonable. It seems that the RA range isn't right, and that there is no clipping box. Can you summarize what exactly you tried to do and what failed?

**Comment by rainwoodman** [comment]
The first figure I did something like set_xlim(350, 40). The second figure I did set_xlim(-40, 40). 

I am trying to map the standard matplotlib jargons to aea projection.
- set_xlim: set ra lim which maps to axes coord 0, 1
- set_ylim: set dec lim which maps to axes coord 0, 1
- subplot(... **kwargs) sets ra_0, dec_0, dec_1, dec_2 for the transformation
- position  in RA DEC are transformed by transData
- meridians and parallels and labels are set by major ticks (not quite there yet)
- lines given as RA,DEC control knobs are 'recursively' refined /interpolated to follow the geometry
- edges of the plotted area (between xlim/ylim) are drawn with matplotlib 'spine's
- clipping with a path the overlaps with the spine (unimplemented)
- polygons given as RA,DEC vertices are correctly plotted to the map as long as all vertices are consistent: 359 to 1 will cover almost the entire sky, vs -1 to 1 covers a thin slice

**Comment by rainwoodman** [comment]
What I have been failing is to build a clean intuition on the corner cases. The 2pi Periodicity is killing my brain!

**Comment by pmelchior** [comment]
I see. The standard treatment is to map everything outside of -180 .. 180 deg onto that range, like here: https://github.com/pmelchior/skymapper/blob/master/skymapper.py#L55

**Comment by rainwoodman** [comment]
then when a patch is near 180 deg from ra_0, the right side edge will be wrapped to -180, and the patch is broken.

Since TransProjection has no idea about 'neighbourhood' or the other vertices of the patch, it seems to be inherently broken to insist wrapping in TransProjection.

I would propose to provide a set of helper functions for wrapping points and vertices properly into the requested RA/DEC range of an axes before calling plot or creating a collection.

**Comment by rainwoodman** [comment]
This is the lastest result.

the RA of the input vertices is between 0~ 360. xlim is (-80, 360-80). those outside (-80, 360-80) are clipped.

index

**Comment by pmelchior** [comment]
1) about the wrapped patches: I get your point that these are no simple polygons anymore. In fact a single polygon that crosses +-180 will have to be split into 2. A wrapper might be required. I'll have a look at how the folks who deal with shape files for a living (like https://github.com/mbloch/mapshaper) deal with this problem.
2) for your latest example: Is green the clippath (it looks like the full 360 deg)? Since your vertices span the entire 360 deg, one would expect them to wrap around. However, I find the xlim(-80, 360-80) rather odd. I'd expect that there is some ra_0 between 0 and 360, which tells you what x=0 is, and xlim will strictly be inside that range, too.

**Comment by rainwoodman** [comment]
Yes. Green was the clip path. ra_0 was at 100. 

What you mean x=0 -- which coordinate are you referring to?

**Comment by rainwoodman** [comment]
- coordinate system.

**Comment by pmelchior** [comment]
I meant that (ra_0, dec_0) maps onto the origin (0,0) of a ordinary right-handed coordinate frame. I called that x=0, y=0 before, which isn't ideal given matplotlib's terminology. You've clearly implemented ra_0 correctly, so my question is: why use xlim(-80, 360-80) instead of (0, 360)?

**Comment by rainwoodman** [comment]
Two motivation:
1. For testing purpose. 
2. To not breaking up the footprint into two pieces. e.g. if the footprint is from -40 to 180.

**Comment by pmelchior** [comment]
Yes, 2. is a good argument. So the question remains where did the patches go?

**Comment by rainwoodman** [comment]
The have been clipped by the clip path on the left.

I think we shall define a new artist type HealpixMap(), subclass of PolyCollection that automatically cull and break the patches to the RA limit of the axes, when 'draw' is called. And we shall provide ax.healpixshow() method that adds this artist to the axes.

Of course we need to first figure out how to cull and break the patches into the RA limit.

**Comment by pmelchior** [comment]
Understood. I suggest to merge what you have now, and then open a new fork for dealing with Healpix (or other) polygons. Folks with point data, which should be safe, could then already use the new transformation for their exploratory work.

**Comment by pmelchior** [comment]
For now, users shouldn't go crazy with the clipping either, but that is a lot of progress already.

**Comment by pmelchior** [comment]
FYI: I found this about cutting polygons along the +-180 line: http://bl.ocks.org/mbostock/3788999
The trick is to pre-cut the polygons, and to make sure that the projections knows how to deal with that.

**Comment by rainwoodman** [comment]
Sure. i don't see how they do the cutting yet. But I do like they use greek letters for variable names.

**Comment by rainwoodman** [comment]
Talking about healpix, have you thought about getting rid of the healpy dependency? 

There is so much more than what we use in healpy.

**Comment by pmelchior** [comment]
I've been having issues with at least the clip box when I try the run the minimal example you posted above:

```
import matplotlib.pyplot as plt
import numpy as np
import aea_projection
fig = plt.figure(figsize=(6, 6))

ra = np.random.uniform(size=10000, low=0, high=360)
dec = np.random.uniform(size=10000, low=-90, high=90)

ax = fig.add_subplot(111, aspect='equal', projection="aea", dec_1=-20., dec_2=30., ra_0=180, dec_0=0.)
ax.plot(ra, dec, '.')
ax.grid()
plt.show()
```

The result is this, which shifted to the very left of the range:

minimal

Also, the clipbox appear correct on the left and bottom, but I still get cursor coordinates in the top area above the solid line.

**Comment by pmelchior** [comment]
Regarding healpy: In DES, we tend to use it a lot. In fact, I'll add a wrapper to directly display healpy-generated maps. I understand that others may not want or need it. But the import is isolated in only those few functions, which don't need to be used.

That being said, if you have another fast way to do the bean-counting, we could in principle replace the functionality in `getCountAtLocations` with it. For me that sounds like more trouble than it's worth.

**Comment by rainwoodman** [comment]
Try this:

```
    ax = fig.add_subplot(111, aspect='equal', projection="aea", dec_1=-20., dec_2=30., ra_0=180, dec_0=0.)
    ax.set_xlim(0, 360)
    ax.set_ylim(-70, 70)
    ax.plot(ra, dec, '.')
    ax.grid()
```

There must be a bug when set_xlim is not called. The state machine in matplotlib is not very well understood by me, I have to admit. A fix shall be easy.

There is a c implementation of healpix (chealpix) which provides no 'additional values' as by healpy. 

I did write a python binding of chealpix several years ago. If you are open to adding in a cython extension module it may be time to revisit that binding. 

Otherwise, I wonder if it is possible to do a pure python version of healpix for the minimal functionality that we use -- the algorithm is not very complicated if we do not insist on being bit-wise identical.

**Comment by pmelchior** [comment]
Merged your new PR, which indeed fixes the axes setup. Thanks.

As for healpy: I think it's either pure python for the pixelation functions on the sphere, or we stick to healpy. I agree that it shouldn't be hard, but it doesn't have high priority for me either.
